Use stochastic restart patterns on rerun

WalterKolczynski-NOAA opened the pull request:
# Description
   
The stochastic pattern restart files were not being copied into the input directory when restarting the model after a segment/failure. These files are now copied in and the stochini flag set to `.true.` in the namelist on a rerun.
    
The files are NOT copied in for non-rerun warm starts.

Also removes the restriction that stochastic physics cannot be run on member 0, as this is desired down the line. Additional settings are added to the fcst and efcs configs to retain the current behavior.

A bug was also discovered and corrected during this work. The stage, forecast, and archive job all assumed that `ca_data` tile files are always present, but these are only created when cellular automata is on. Now `ca_data` files are handled if CA is on. To implement this change, the `DO_CA` setting had to be moved from the forecast configs to base so it is available to the stage_ic and archive jobs.

Resolves #2937

# Type of change
- [x] Bug fix (fixes something broken)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO
- Does this change require an update to any of the following submodules? NO

# How has this been tested?
Multi-segment test on Hercules

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have documented my code, including function, input, and output descriptions
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] This change is covered by an existing CI test or a new one has been added
- [x] Any new scripts have been added to the .github/CODEOWNERS file with owners
- [x] I have made corresponding changes to the system documentation if necessary

WalterKolczynski-NOAA committed:
Use stochastic restart patterns on rerun

The stochastic pattern restart files were not being copied into the
input directory when restarting the model after a segment/failure.
These files are now copied in. Additional changes appear to be
needed for them to be used by UFS.

The files are NOT copied in for non-rerun warm starts.

Refs #2937

WalterKolczynski-NOAA committed:
Set stochini to true to use stoch restart patterns

Adds the namelist flag so the stochastic pattern restart files
get read when doing a RERUN.

Refs #2937

WalterKolczynski-NOAA committed:
Remove commented out if and unindent

aerorahul reviewed (dismissed):
looks good to me.  No comments.
Please invite review from @pjpegion @NeilBarton-NOAA

NeilBarton-NOAA commented on ush/forecast_postdet.sh:
```diff
@@ -47,7 +47,19 @@ FV3_postdet() {
     done
 
-    if [[ "${RERUN}" != "YES" ]]; then
+    if [[ "${RERUN}" == "YES" ]]; then
```
@bingfu-NOAA @EricSinsky-NOAA Is there a time that we want to start the model with these files? i.e., use them in more ways than a RERUN?

bingfu-NOAA replied:
For 00Z, we will stop at 16-day forecast, then warm start the model to run 16-48 days forecast segment. We will need those files at that time besides a RERUN.

WalterKolczynski-NOAA replied:
Subsequent segments count as RERUNs inside the system.

bingfu-NOAA commented on parm/config/gefs/config.fcst:
```diff
@@ -227,11 +227,13 @@ case ${imp_physics} in
+export DO_SKEB="NO"
+export DO_SHUM="NO"
+export DO_LAND_PERT="NO"
+export DO_CA="NO"
```
is this config file being used by mem000 in ensemble?, if so, then DO_CA should be YES.

WalterKolczynski-NOAA replied:
Done

bingfu-NOAA commented on parm/config/gfs/config.fcst:
```diff
@@ -257,14 +257,13 @@ case ${imp_physics} in
+export DO_SKEB="NO"
+export DO_SHUM="NO"
+export DO_LAND_PERT="NO"
+export DO_CA="NO"
```
GFS is using CA, so DO_CA should be YES.

WalterKolczynski-NOAA replied:
Done

WalterKolczynski-NOAA committed:
Only copy ca_data restart files when CA is on

The forecast job would attempt to copy the ca_data tiled restart
files even when cellular automata was off. Now it only copies them
if cellular automata is on.

WalterKolczynski-NOAA committed:
Update stage and archive to conditionally handle ca_data

The stage and archive jobs were assuming that the ca_data restart
files will always be present. However, these are only produced
when cellular automata is run. So now they are conditionally
handled based on `DO_CA`.

To accomodate this, the `DO_CA` setting had to be moved out of
`config.fcst`/`config.efcs` and into `config.base`.

WalterKolczynski-NOAA committed:
Add missing comma in stage ic script

WalterKolczynski-NOAA committed:
Restore current CA setting

Reverts cellular automata to be on for all situations to match
current develop

CatherineThomas-NOAA commented on parm/config/gfs/config.base:
```diff
@@ -448,6 +448,15 @@ export INCVARS_EFOLD="5"
 export binary_diag=".false."
 
+# Cellular automata
+case ${RUN} in
```
I think this is leftover from mixed `DO_CA` settings, but is the `case` section necessary now?

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/stoch_restart

WalterKolczynski-NOAA committed:
Remove RUN dependence for DO_CA in GFS

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/stoch_restart

WalterKolczynski-NOAA commented:
This works for me manually. I'm going to try again and inspect it as it runs.

WalterKolczynski-NOAA committed:
Turn off ocean SPPT above 1p00

The MOM_input template for 5p00 does not contain the settings for
ocean SPPT or EPBL, so don't turn them on to avoid workflow from
trying to copy non-existent files.

Also stop ignoring a dirty UFS model directory to avoid masking
changes.

bingfu-NOAA commented on parm/stage/atmosphere_warm.yaml.j2:
```diff
@@ -15,9 +15,12 @@ atmosphere_warm:
         {% for ftype in ["coupler.res", "fv_core.res.nc"] %}
         - ["{{ ICSDIR }}/{{ COMOUT_ATMOS_RESTART_PREV_MEM | relpath(ROTDIR) }}/{{ m_prefix }}.{{ ftype }}", "{{ COMOUT_ATMOS_RESTART_PREV_MEM }}"]
         {% endfor %}
-        {% for ftype in ["ca_data", "fv_core.res", "fv_srf_wnd.res", "fv_tracer.res", "phy_data", "sfc_data"] %}
```
What is the reason to remove ca_data?

WalterKolczynski-NOAA replied:
It's not removed, it is just only added to the list when CA is on.

bingfu-NOAA commented on ush/forecast_predet.sh:
```diff
@@ -377,41 +396,38 @@ FV3_predet(){
-    fi
+  if [[ "${DO_CA:-}" == "YES" ]]; then
+    do_ca=".true."
+    ISEED_CA=$(( (base_seed + 18) % 2147483647 ))
```
As CA is being used by both GFS/GEFS, probably we can just use
ISEED_CA=$(( (base_seed) % 2147483647 ))

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/stoch_restart

WalterKolczynski-NOAA committed:
Fix turning off ocean stoch phys above 1p00

DavidHuber-NOAA commented on parm/config/gefs/config.base:
```diff
@@ -287,10 +287,17 @@ export DO_MERGENSST="NO"
 
+# cellular automata
+if (( ENSMEM == 0 )); then
+    export DO_CA="YES"
+else
+    export DO_CA="YES"
```
Should one of these be `NO`?

WalterKolczynski-NOAA replied:
Nope, at least not by default right now.

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/stoch_restart

WalterKolczynski-NOAA commented:
CI Tests set up to run in /lfs/h2/emc/ptmp/walter.kolczynski/PR/PR_3077/RUNTESTS on WCOSS

WalterKolczynski-NOAA commented:
Got all the way to the last cycle of the extended test and then died. Investigating.

WalterKolczynski-NOAA commented:
Looks like stmp filled on WCOSS

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/stoch_restart

WalterKolczynski-NOAA committed:
Add DO_CA to earc keys

WalterKolczynski-NOAA committed:
Fix earc path for ca_data

KateFriedman-NOAA reviewed (approved):
I can't comment on the science updates in this PR but the workflow changes look good. Thanks @WalterKolczynski-NOAA !
